tests/integ/scala/test_dataframe_join_suite.py

```diff
@@ -1143,3 +1143,19 @@ def test_join_diamond_shape_workaround(session):
     df1_converted = df1.select(df1["a"])
     df5 = df1_converted.join(df4, df1_converted["a"] == df4["a"])
     Utils.check_answer(df5, [Row(1, 1)])
+
+
+def test_dataframe_basic_diamond_shaped_join(session):
+    df1 = session.create_dataframe([[1, 2], [3, 4], [5, 6]], schema=["a", "b"])
+    df2 = df1.filter(col("a") > 1).with_column("c", lit(7))
+    assert df1.a._expression.expr_id != df2.a._expression.expr_id
+
```

Comment:
why are we testing `expr_id`s are not the same. Do it affect dataframes in weird ways if they are same? Can we test that the subsequent side effects of `expr_id`s being the same are tested instead?

Reply:
If they are the same we won't be able to perform the join. https://github.com/snowflakedb/snowpark-python/blob/df013e58b6aa575f36700aedb094b355109871b2/src/snowflake/snowpark/_internal/analyzer/analyzer.py#L732 in the analyzer counts duplicate expression id's in the resolved children and delete the <expr_id, alias name> entries in the analyzer's map to prevent ambiguous references, but then analyzer will translate df1.a to the "A", where the actual name of df1.a is suffixed. So the generated query will fail.